runtime/compact.c

```diff
@@ -483,41 +474,26 @@ void caml_compact_heap_maybe (void)
     return;
 #endif
 
-  fw = 3.0 * caml_fl_cur_wsz - 2.0 * caml_fl_wsz_at_phase_change;
-  if (fw < 0) fw = caml_fl_cur_wsz;
+  if (previous_overhead >= caml_percent_max){
+    double current_overhead;
 
-  if (fw >= Caml_state->stat_heap_wsz){
-    fp = 1000000.0;
-  }else{
-    fp = 100.0 * fw / (Caml_state->stat_heap_wsz - fw);
-    if (fp > 1000000.0) fp = 1000000.0;
-  }
-  caml_gc_message (0x200, "FL size at phase change = %"
-                          ARCH_INTNAT_PRINTF_FORMAT "u words\n",
-                   (uintnat) caml_fl_wsz_at_phase_change);
-  caml_gc_message (0x200, "FL current size = %"
-                          ARCH_INTNAT_PRINTF_FORMAT "u words\n",
-                   (uintnat) caml_fl_cur_wsz);
-  caml_gc_message (0x200, "Estimated overhead = %"
-                          ARCH_INTNAT_PRINTF_FORMAT "u%%\n",
-                   (uintnat) fp);
-  if (fp >= caml_percent_max){
     caml_gc_message (0x200, "Automatic compaction triggered.\n");
     caml_empty_minor_heap ();  /* minor heap must be empty for compaction */
     caml_gc_message
       (0x1, "Finishing major GC cycle (triggered by compaction)\n");
     caml_finish_major_cycle ();
     ++ Caml_state->stat_forced_major_collections;
 
-    fw = caml_fl_cur_wsz;
-    fp = 100.0 * fw / (Caml_state->stat_heap_wsz - fw);
-    caml_gc_message (0x200, "Measured overhead: %"
+    /* Note: There is no floating garbage because we just did a complete
+       major cycle*/
+    current_overhead =
+      100.0 * caml_fl_cur_wsz / (Caml_state->stat_heap_wsz - caml_fl_cur_wsz);
```

Comment:
Below you have:

```
      previous_overhead =
        100.0 * (heap_wsz_at_cycle_start - marked_words) / marked_words;
```

I believe that both are correct, but it would be nice to make the two definitions easier to compare with simple refactorings.

```
{
  free_words = caml_fl_cur_wsz;
  live_words = Caml_state->stat_heap_wsz - caml_fl_cur_wsz;
  current_overhead = 100 * free_words / live_words;
}

{
  free_words = heap_wsz_at_cycle_start - marked_words;
  live_words = marked_words;
  current_overhead = 100 * free_words / live_words;
}
```

Reply:
Maybe a silly question: in the version below you defend against division-by-zero but not here. Is there a reason for the difference? (Division-by-zero occurs when no words are live anymore, I don't know how it could happen in practice.)